README
Full featured HTTP SMS service for 2n VoiceBlue (Next) GSM VoIP adapters with sending and receivig capabilities.
Telnet or serial communications are available.

The service periodically checks the incoming SMS-es on the VoiceBlue adapter then stores in a local SQLite database.
It can communicate via telnet, you can switch telnet off with USE_TELNET = False/True. 
If USE_TELNET = False, it will try to connect ONLY on serial (ttyUSB) ports (older VoiceBlue devices, but telnet preferred).

The HTTP REST service listens on a defined HTTP port (default 8080) and controllable via simple POST/GET calls. (In case of GET call the message should be URL encoded.) The HTTP responses from the service are simple XMLs for easy to use and parse.
For details, control functions, output XML templates, please read the README document.
There is a basic Apache+PHP WEB frontend available to control the service.

VoiceBlue AT commands reference:
https://wiki.2n.cz/display/VBN30EN/4.2+AT+Interface

DEPENDENCIES:
- Python > 2.7.x
- Python-Messaging https://github.com/pmarti/python-messaging

PREFERRED way to install Python-messaging module:
cd ~
git clone https://github.com/pmarti/python-messaging.git
cd python-messaging
sudo python setup.py install 

OPTIONAL:
For serial communication:
- Python-serial 2.5 > http://sourceforge.net/projects/pyserial/files/pyserial/2.5/pyserial-2.5.tar.gz/download

BASIC USAGE EXAMPLE:
0.) edit config variables below: # CONFIG START section
1.) start HTTP server on port: ./[program name] [server port] (default:8080) (root user requitred ONLY for serial communictaions)
2.) sending SMS: use a HTTP browser to call the server via GET method: 
	http://[host address:port]/?rcpt=[full tel. number of recipient]&msg=urlencode([155 character sms])&smsc=[our telco's sms center full number (with + and country code, eg: +36...)]
3.) receiving SMS messages: http://[host address:port]/?action=listall

Licensed under GPL v2: http://www.gnu.org/licenses/gpl-2.0.html 

------------------------------------------
DETAILED CONFIG
------------------------------------------

For configuration, please edit the 2nVoiceBlueSMSServer.py file "CONFIG section."

SIMCARDSINUSE:	how many SIM card installed on yout device. (0,1,2,3)
SQLITE_PATH:	the local SQLite database path, the server will creates it if not exist. The sent and received SMS-es stored here. 
				IMPORTANT: the server deletes the SMS messages from the SIM cards after it fetched and stored in teh SQLite database.
LOGFILE:		the detailed log output of the server
LOG_TO_LOGFILE: 	if you need ONLY print to CLI, use False here
SERPORT:		basic serial USB path root, without number! of the device, at startup, the server try to connect from ttyUSB[0] to ttyUSB[10]
SERBAUD:		baud rate for 2n device, fixed : 921600, do not change it!
PORT:			the HTTP server default port to listen on
SMSC:			default telco's SMSC number. It can be pass via HTTP GET also, but you can set it up. Can be fetched from the SMS, Message settings, or from the Internet.
SIMPOLLINTERVAL:	the server periodically polls the SIM cards in the device, fetches the SMS messages then stores in the local db, the poll interval in seconds
#telnet settings
USE_TELNET:			True: use telnet to connect VB device, it is the default, False: try serial ports
TELNET_HOST:		telnet hostname or IP of VoiceBlue modem
TELNET_PORT:		telnet port
TELNET_USER:		Voice Blue modem user
TELNET_PASSWORD		VoiceBlue modem password
TELNET_TIMEOUT:		the telnet command timeout

USE_GSM_MODEM: True/False, workaround for CMS: 500 internal error, use GSM modem directly, workaround for Voice Blue Next devices, keep it True

------------------------------------------
HTTP COMMUNICATION, GET/POST CALLS AND RESPONSES
------------------------------------------

--------------------------
SMS SENDING via POST/GET:
--------------------------
example: http://127.0.0.1/?rcpt=+36120300001234&msg=tryitout

rcpt:		recipient phone number
msg:		SMS message body, max 155 chars, accents removed. To avoid encoding problems, it should be urlencoded in case of GET.
smsc:		SIM's telco's SMS center number, not required if given at the config above.

Response XML:
-------------
<response type="send">
	<status>[OK|ERROR]</status>
	<date>[datetime]</date> <-- datetime format: YYYY-MM-DD hh:mm:ss -->
	<id>[individual SMS ID]</id>
</response>

--------------------------
SMS FETCH via POST/GET:
--------------------------
example: http://127.0.0.1/?action=listall

action:		listall		(list all plus deleted SMS messages stored locally)
			list		(list only not deleted SMS messages)
			listdel		(list deleted SMS)
			listrec		(default / list only received - incoming - SMS messages)
			listrecdel	(list deleted received messages)
			listsent	(list sent  - outgoing - messages)
			listsentdel	(list received deleted messages)
			listqueue	(list queued messages for sending - not sent yet)
			listerr		(list of messages with sending errors occured)

Response XML:
-------------
<response type="receive">
	<sms>
		<id>[SHA1 hash id of SMS]</id>
		<smsdate>[sent/received datetime]</smsdate>
		<lastdate>[datetime of message modification - queued -> sent, error]</lastdate>
		<cardnum>"[SIM card number: 0|1|2|3]</cardnum>
		<fromnum>[sender number - only incoming messages | None]</fromnum>
		<tonum>[recipient's number - only sent messages | None]</tonum>
		<status>[queued | sent  | senderror | received]</status>
		<deleted>[yes | no]</deleted>
		<msg><![CDATA[ - message body - ]]></msg>
	</sms>
</response>

--------------------------
SMS DELETE via POST/GET:
--------------------------
example: http://127.0.0.1/?action=delete&id=[SMS id - hash]

action:		delete
id:			SHA1 hash (40 chars long ID)

Response XML:
-------------
<response type="delete">
	<status>[OK|ERROR]</status>
	<date>[datetime]</date> <-- datetime format: YYYY-MM-DD hh:mm:ss -->
	<id>[individual SMS ID]</id>
</response>

--------------------------
TRUNCATE DELETED SMS FROM DB via POST/GET:
--------------------------
example: http://127.0.0.1/?action=truncate

action:		truncate

Response XML:
-------------
<response type="truncate">
	<status>[OK|ERROR]</status>
	<date>[datetime]</date> <-- datetime format: YYYY-MM-DD hh:mm:ss -->
	<pcs>[number of truncated messages]</pcs>
</response>


--------------------------
INSUFFICIENT RESPONSE:
--------------------------
<response type="error">
	<error>insufficient request</error>
</response>

For install instructions please check INSTALL file.


CHANGELOG
3.10 (2014-10-02)
-----------------
- fixed some unicode errors
- workaround for Voice Blue Next SMS sending by using GSM modems directly
- fixed Voice Blue next SMS receiving
- internal GSM modems status
- random outgoing SMS sending over accessible modems
- fix telnet init and connection keepalive
- speed up operations by re-using telnet connection if available

3.0 (2014-10-01)
-----------------
- added telnet connection capabilities

2.0 (2013-01-30)
-----------------
- initial github release
- only ttyUSB or serial connection available
- basic web interface (PHP)



var/www/2nsms/index.php
<?php

/* 
 * Basic WEB frontend for 2nVoiceBlueSMSServer.py SMS server
 * 
 *  __author__ = "Tamas TOBI <tamas.tobi@gmail.com>, Hungary"
 * __copyright__ = "Copyright (C) Tamas Tobi 2012, License: GPL v2, http://www.gnu.org/licenses/gpl-2.0.html"
 * 
 * Dependencies: PHP5.x, Webserver: preferrable Apache2.x
*/

////////////////////////////////////////////////////
// CONFIG
define('WEBLOGPATH',"/var/www/2nsms/smssend.log"); 	//Logging path, www-data user can be write here
define('SMSSERVERIP','127.0.0.1');					//The remote IP where the 2nVoiceBlueSMSServer runs
define('SMSSERVERPORT','8080');						//2nVoiceBlueSMSServer remote PORT
// CONFIG END
////////////////////////////////////////////////////


header('Content-Type: text/html; charset=utf-8');
@ini_set('zlib.output_compression',0);
@ini_set('implicit_flush',1);
@ob_end_clean();
set_time_limit(0);

function elog($msg){
	error_log(date('Y-m-d H:i:s')."\t".$msg."\n",3,WEBLOGPATH);
}

function vphone($inp){
	$v='0123456789';
	$var = str_split($v); 

	$o='';
	$inp=trim((string)$inp); 

	$inp=str_replace("'","",$inp);
	$inp=str_replace('"',"",$inp);
	$inp=str_replace(' ',"",$inp);
	$inp=str_replace(',',"",$inp);
	$inp=str_replace(';',"",$inp);
	$inpar = str_split($inp);
	foreach($inpar as $i){
		if (in_array($i, $var)){
		  $o.=$i;
		}
	}
	return $o;
}

function Selector($order_name,$field_name,$title,$selected=false) {
	$rtrn = "<option";
	if ((isset($_GET[$order_name]) && $_GET[$order_name] == $field_name) || (isset($_POST[$order_name]) && $_POST[$order_name] == $field_name) ||  $selected) {
		$rtrn .=' selected="selected" ';
	} else {
		$rtrn .= ' ';
	}
	$rtrn .= 'value="'.$field_name.'">';
	$rtrn .= $title;
	$rtrn .= '</option>';
	return $rtrn;
}

//sms database actions
$action = (isset($_GET['action']) && !empty($_GET['action'])) ? trim($_GET['action']) : "listrec";
$mod = (isset($_GET['mod']) && !empty($_GET['mod'])) ? trim($_GET['mod']) : "";
$_GET['action'] = $action;

$smslist = false;

if ($mod == "delete"){
	$msgid = $_GET['id'];
	$smslist = file_get_contents("http://".SMSSERVERIP.":".SMSSERVERPORT."/?action=".$mod."&id=".$msgid);
	header("Location: ./?action=".$action);
}
if ($mod == "truncate"){
	$smslist = file_get_contents("http://".SMSSERVERIP.":".SMSSERVERPORT."/?action=".$mod);
	header("Location: ./?action=".$action);
}

$smslist = file_get_contents("http://".SMSSERVERIP.":".SMSSERVERPORT."/?action=".$action);
if ($smslist && !empty($smslist)){
	$xml = simplexml_load_string($smslist);

	$smstable='<table border="0" cellpadding="2" width="825px">';
	$smstable.='<tr bgcolor="#afafaf" ><th>Date</th><th>From</th><th>To</th><th>Status</th><th>SIM</th><th>Message</th><th>Del</th></tr>';
	$i=0;
	foreach($xml->children() as $sms){
		if ($i%2){
			$smstable.= '<tr bgcolor="fafafa">';
		} else {
			$smstable.= '<tr bgcolor="e1e1e1">';
		}
		$smstable.= '<td width="120px">'.$sms->smsdate.'</td>';
		$smstable.= '<td>'.$sms->fromnum.'</td>';
		$smstable.= '<td>'.$sms->tonum.'</td>';
		$smstable.= '<td>'.$sms->status.'</td>';
		$smstable.= '<td>'.$sms->cardnum.'</td>';
		$smstable.= '<td>'.$sms->msg.'</td>';
		$smstable.= '<td><a href="./?action='.$action.'&id='.$sms->id.'&mod=delete">['.$sms->deleted.']</a></td>';
		$smstable.= '</tr>';
		$i++;
	}
	$smstable.='</table>';
}

?>
<!DOCTYPE html>
<html lang="en-US">
	<head>
		<meta charset="utf-8">
		<title>2n VoiceBlue SMS Server - Basic Web Interface</title>
		<style>
			body {
				font-size: 12px;
				font-family: Arial, Helvetica, sans-serif;
				line-height: 20px;
				color: #333333;
			}
			
			a {
				color: #blue;
				text-decoration: underline;
			}
			
			.overview {
				background: #FFEC9D;
				padding: 10px;
				width: 800px;
				border: 1px solid #CCCCCC;
			}
			
			.originalTextareaInfo {
				font-size: 12px;
				color: #000000;
				font-family: Tahoma, sans-serif;
				text-align: right
			}
			
			.warningTextareaInfo {
				font-size: 12px;
				color: #FF0000;
				font-family: Tahoma, sans-serif;
				text-align: right
			}
			
		</style>
		<script src="./jquery.min.js" type="text/javascript"></script>
		<script src="./jquery.textareaCounter.plugin.js" type="text/javascript"></script>
		<script type="text/javascript">
			var info;
			$(document).ready(function(){
				var options2 = {
						'maxCharacterSize': 155,
						'originalStyle': 'originalTextareaInfo',
						'warningStyle' : 'warningTextareaInfo',
						'warningNumber': 150,
						'displayFormat' : '#input/#max'
				};
				$('#msgTextarea').textareaCount(options2);
			});
		</script>
	</head>
	<body>
	<div align="center">
<?php

//sms sending 
if (isset($_POST['numTextarea']) 
	&& !empty($_POST['numTextarea']) 
	&& strlen(trim($_POST['numTextarea'])) > 3
	&& isset($_POST['msgTextarea'])){

	echo '<div class="overview">
		SMS sending result
		</div>';
	flush();
	$numary=explode("\n",$_POST['numTextarea']);
	$msg=trim($_POST['msgTextarea']);
	$msg=str_replace(array("\r", "\n", "\r\n"), ' ', $msg);
	$msg=trim($msg);
	elog("---------------");
	elog($_SERVER['REMOTE_ADDR'].' / SMS sending: '.str_replace(array("\r", "\n", "\r\n"), ',',$_POST['numTextarea']).": ".$msg);
	echo "Message: ", $msg,"<hr>";
	$i=1;
	foreach($numary as $phone){
		$phone=trim($phone);
		if (!$phone){
			continue;
		}
		$vphone=vphone($phone);
		if (!is_numeric($vphone)){
			echo '<font color="red"><b>Invalid number: ',$phone,' => ',$vphone,' !!! </b></font><br>';
			flush();
			sleep(1);
			continue;
		}
		$handle = file_get_contents("http://".SMSSERVERIP.":".SMSSERVERPORT."/?rcpt=".$vphone."&msg=".urlencode($msg)."");
		elog($handle);
		echo $handle,"<br>";
		flush();
		sleep(1);
		$i++;
	}
	elog("---------------");
	echo 'Sent: ',($i-1)," pcs";
	sleep(2);
}
?>
<div class="overview">
2N VoiceBlue - basic SMS frontend
</div>
<form action="./" method="post">
	<table border="0" cellpadding="10">
		<tr><td valign="bottom"><b>Recipient numbers (per row):</b></td><td valign="bottom"><b>Message (max 155 char):</b></td></tr>
		<tr><td valign="top"><textarea id="numTextarea" name="numTextarea"  cols="25" rows="10"></textarea></td><td valign="top"><textarea id="msgTextarea" name="msgTextarea" cols="25" rows="10"></textarea></td></tr>
		<tr><td colspan="2" valign="top"><input type="submit" value="Send SMS" /></td></tr>
	</table>
</form>
<div class="overview">SMS database</div>
<br />
<form action="./" method="get">
<b>Filter:</b> <select id="action" name="action" onChange="form.submit();">';
<?php 
echo Selector("action","listrec","Received SMS");
echo Selector("action","listrecdel","Received SMS - show deleted");
echo Selector("action","listall","All SMS (sent+received)");
echo Selector("action","list","All SMS (sent+received) - show not deleted");
echo Selector("action","listdel","All SMS (sent+received) - show deleted");
echo Selector("action","listsent","Sent SMS");
echo Selector("action","listsentdel","Sent SMS - show deleted");
echo Selector("action","listqueue","SMS queued for sending");
echo Selector("action","listerr","SMS errors");
?>
</select> <button type="button" onclick="window.location.href='./?action=<?=$action?>'">Page Refresh</button> <button type="button" onclick="window.location.href='./?mod=truncate&action=<?=$action?>'">Truncate Deleted</button>
</form>
<br />
<?=$smstable?>
</div>
</body>
</html>
